core/celo_multi_gaspool.go

```diff
@@ -0,0 +1,79 @@
+package core
+
+import "github.com/ethereum/go-ethereum/common"
+
+type FeeCurrency = common.Address
+
+// MultiGasPool tracks the amount of gas available during execution
+// of the transactions in a block per fee currency. The zero value is a pool
+// with zero gas available.
+type MultiGasPool struct {
+	pools       map[FeeCurrency]*GasPool
+	defaultPool *GasPool
+
+	blockGasLimit uint64
+	defaultLimit  float64
+}
+
+type FeeCurrencyLimitMapping = map[FeeCurrency]float64
+
+// NewMultiGasPool creates a multi-fee currency gas pool and a default fallback
+// pool for CELO
+func NewMultiGasPool(
+	blockGasLimit uint64,
+	defaultLimit float64,
+	limitsMapping FeeCurrencyLimitMapping,
```

Comment:
Since we dropped the whitelist parameter here, the multipool can accept arbitrary mappings now and technically is unbound. Probably not a big deal but just wondering, if that's intentional and what's the reasoning.

Reply:
When the `NewEVMBlockContext` function [is called](https://github.com/celo-org/op-geth/blob/celo4/miner/worker.go#L1062), it makes a [call to fetch](https://github.com/celo-org/op-geth/blob/celo4/core/celo_evm.go#L20) exchange rates for fee currencies. This exchange rate map is then [used](https://github.com/celo-org/op-geth/blob/celo4/common/celo_types.go#L13) as a way to check if a currency is whitelisted [throughout](https://github.com/celo-org/op-geth/blob/celo4/core/state_transition.go#L419) the code. I thought we could reuse it here as well. I understand that the TX pool does the check but I think we should re-validate.
@palango what do you think?

Reply:
The advantage of removing the check I see is that a node operator could add limits for fee currencies before they are whitelisted. But I'm not sure if that's a strong point.

I see more value in having the configuration checked, so that the node operator doesn't accidentally pass a wrong address and is never alerted that the address is not a fee currency.

Reply:
> I see more value in having the configuration checked, so that the node operator doesn't accidentally pass a wrong address and is never alerted that the address is not a fee currency.

You mean a configuration-validation step that emits a warning (or fails the node?) upon node-initialization?
I think this would be a good addition, I can create a follow up issue if we decide on that feature.

Reply:
I changed the implementation so that it copies the exact behavior as https://github.com/celo-org/celo-blockchain/commit/dc45bdc00. 
The exchange rates are now taken from the block context @carterqw2  mentioned and attached to the worker environment.

@carterqw2 Could you review the changes again?

Reply:
Looks good overall, left one tiny comment.